### PR TITLE
modify to have a list of precompiled headers

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: build_functions/CMakeLists.txt [hunk 1]
```diff
@@ -25,36 +33,82 @@ function(build_lib whatIsBuilding solution_folder)
...
+    if(arg_MOCK_PRECOMPILE_HEADERS OR arg_NO_MOCK_PRECOMPILE_HEADERS)
+
+        string(TOUPPER ${whatIsBuilding} LIBRARY_NAME_UPPER)
+
+        set(MOCKS_HEADER_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.h)
+        set(MOCKS_SOURCE_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.c)
+
+        set(MOCKS_HEADER_CONTENT "// Copyright (c) Microsoft. All rights reserved.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// Licensed under the MIT license. See LICENSE file in the project root for full license information.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// THIS IS A GENERATED FILE, DO NOT EDIT. It was generated using function build_lib from CMakeLists.txt in testrunnerswitcher.\n")
```
Comment (dcristoloveanu) [inline build_functions/CMakeLists.txt:65]: >nnerswitche [](http://example.com/codeflow?start=170&length=11)

Consider putting the link to the CMakeList.Txt :-) Easy to find.

File: build_functions/CMakeLists.txt [hunk 2]
```diff
@@ -25,36 +33,82 @@ function(build_lib whatIsBuilding solution_folder)
...
+        endforeach()
+        set(MOCKS_SOURCE_CONTENT "${MOCKS_SOURCE_CONTENT}#undef ENABLE_MOCKS\n")
+
+        file(WRITE ${MOCKS_SOURCE_FILE} ${MOCKS_SOURCE_CONTENT})
+
+        set_source_files_properties(${MOCKS_SOURCE_FILE} PROPERTIES SKIP_PRECOMPILE_HEADERS ON)
+
+        target_precompile_headers(${whatIsBuilding}_lib_${CMAKE_PROJECT_NAME} PRIVATE ${MOCKS_HEADER_FILE})
+        target_sources(${whatIsBuilding}_lib_${CMAKE_PROJECT_NAME} PRIVATE ${MOCKS_SOURCE_FILE})
+    else()
```
Comment (dcristoloveanu) [inline build_functions/CMakeLists.txt:110]: > else() [](http://example.com/codeflow?start=3&length=7)

This else not needed

File: build_functions/CMakeLists.txt [hunk 3]
```diff
@@ -25,36 +33,82 @@ function(build_lib whatIsBuilding solution_folder)
...
+        set(MOCKS_HEADER_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.h)
+        set(MOCKS_SOURCE_FILE ${CMAKE_CURRENT_BINARY_DIR}/${whatIsBuilding}_mocks.c)
+
+        set(MOCKS_HEADER_CONTENT "// Copyright (c) Microsoft. All rights reserved.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// Licensed under the MIT license. See LICENSE file in the project root for full license information.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// THIS IS A GENERATED FILE, DO NOT EDIT. It was generated using function build_lib from CMakeLists.txt in testrunnerswitcher.\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}// To change this file content, edit build_test_artifacts's arguments from CMakeLists.txt\n\n")
+
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}#ifndef ${LIBRARY_NAME_UPPER}_MOCKS_H\n")
+        set(MOCKS_HEADER_CONTENT "${MOCKS_HEADER_CONTENT}#define ${LIBRARY_NAME_UPPER}_MOCKS_H\n\n")
```
Comment (dcristoloveanu) [inline build_functions/CMakeLists.txt:69]: Novel ways to generate code I see :-).

File: build_functions/CMakeLists.txt [hunk 4]
```diff
@@ -3,6 +3,14 @@
 
 set(trsw_internal_dir ${CMAKE_CURRENT_LIST_DIR} CACHE INTERNAL "")
 
+#build_lib produces a lib that can be used with an exe and/or dll for testing
+#ARGN have the format
+#ADDITIONAL_LIBS list of libraries. The libraries is a list consisting of repeating the following
```
Comment (dcristoloveanu) [inline build_functions/CMakeLists.txt:8]: >g [](http://example.com/codeflow?start=96&length=1)

Nit: add a colon. /consider

**Review by dcristoloveanu — APPROVED** [review]
:shipit:

File: build_functions/CMakeLists.txt (hunk 4 above)
Comment (anporumb) [reply build_functions/CMakeLists.txt:8]: added

File: build_functions/CMakeLists.txt (hunk 1 above)
Comment (anporumb) [reply build_functions/CMakeLists.txt:65]: added http link

File: build_functions/CMakeLists.txt (hunk 3 above)
Comment (anporumb) [reply build_functions/CMakeLists.txt:69]: configure_file() was too much for me :(

File: build_functions/CMakeLists.txt (hunk 2 above)
Comment (anporumb) [reply build_functions/CMakeLists.txt:110]: Yeah, printf debuggin artefact leftover

**Review by dcristoloveanu — APPROVED** [review]
:shipit:

**Pull request merged**